### PR TITLE
Implement the default checks for normal events.

### PR DESCRIPTION
**Pull request by NegativeMjark**

### CODE REVIEW

File: eventauth/eventauth.go [hunk 1]
```diff
@@ -109,12 +108,6 @@ func StateNeededForAuth(events []Event) (result StateNeeded) {
 	return
 }
 
-// memberContent is the JSON content of a m.room.member event.
-type memberContent struct {
-	Membership       string          `json:"membership"`
-	ThirdPartyInvite json.RawMessage `json:"third_party_invite"`
-}
```
Comment (NegativeMjark) [inline eventauth/eventauth.go]: Moved to eventcontent.go

File: eventauth/eventauth.go [hunk 2]
```diff
@@ -220,21 +213,6 @@ func createEventAllowed(event Event) error {
...
-	// Split on the first ":" character since the domain can contain ":"
-	// characters.
-	parts := strings.SplitN(id, ":", 2)
-	if len(parts) != 2 {
-		// The ID must have a ":" character.
-		return "", errorf("invalid ID: %q", id)
-	}
-	// Return everything after the first ":" character.
-	return parts[1], nil
-}
```
Comment (NegativeMjark) [inline eventauth/eventauth.go]: Moved to eventcontent.go

File: eventauth/eventcontent.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
+package eventauth
+
+import (
+	"encoding/json"
+	"strconv"
+	"strings"
+)
+
+// creaateContent is the JSON content of a m.room.create event along with
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: creaateContent

File: eventauth/eventcontent.go [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+// the top level keys needed for auth.
+type createContent struct {
+	// We need the domain of the create event when checking federatebility.
+	senderDomain string
+	// We need the roomID to check that events are in the same room as the create event.
+	roomID string
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: Why is this here? I am grateful for the explanation so I know what it is used for, but it seems weird to have this here. Any justification?

File: eventauth/eventcontent.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+	// The "m.federate" flag tells us whether the room can be federated to other servers.
+	Federate *bool `json:"m.federate"`
+	// The creator of the room tells us what the default power levels are.
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: Really? I thought the `creator` was the ID of the user who called `/createRoom`.

File: eventauth/eventcontent.go [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+func (c *createContent) domainAllowed(domain string) error {
+	if domain == c.senderDomain {
+		// If the domain matches the domain of the create event
+		return nil
+	}
+	if c.Federate == nil || *c.Federate {
+		// The m.federate field defaults to true.
+		return nil
+	}
+	return errorf("room is unfederatable")
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: Yucky, I don't like the style of "filter out successes then what's left is an error" because it is really error-prone when compared to "filter out errors then what's left is a success". For example, what happens in this function if `domain != c.senderDomain` and `c.Federate = true`? The domain will be **allowed** even though the sender domain doesn't match!

File: eventauth/eventcontent.go [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+}
+
+// idAllowed checks whether the domain of the ID is allowed in the room.
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: I believe you're being purposefully vague because this will work for all of the Matrix-style IDs, but it doesn't help people understand what they can or cannot reasonably pass into this function. I would prefer it if this explicitly stated room and user IDs, given this will totally work for room aliases, event IDs, etc which is presumably not what your intention is here.

File: eventauth/eventcontent.go (hunk 6 above)
Comment (NegativeMjark) [inline eventauth/eventcontent.go]: This style is superior cause it list the way the event is allowed. Rather than trying to list all the ways the event is *not* allowed which is liable to miss cases.

But I take your point this bit needs better comments.

File: eventauth/eventcontent.go [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+	parts := strings.SplitN(id, ":", 2)
+	if len(parts) != 2 {
+		// The ID must have a ":" character.
+		return "", errorf("invalid ID: %q", id)
+	}
+	// Return everything after the first ":" character.
+	return parts[1], nil
+}
+
+// memberContent is the JSON content of a m.room.member event.
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: Again, I'm struggling to conceptualise what `memberContent` is actually supposed to be representing. It sure isn't what you're stating here, because it's missing a bunch of fields from http://matrix.org/docs/spec/client_server/r0.2.0.html#m-room-member - given this is for *auth checks* I imagine you only care about the fields which are used for auth, so you should mention this in your description.

File: eventauth/eventcontent.go [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+type memberContent struct {
+	Membership       string          `json:"membership"`
+	ThirdPartyInvite json.RawMessage `json:"third_party_invite"`
+}
+
+// Load the member content from the member event for the user ID.
+// Returns an error if there was an error loading the member event or
+// parsing the event content.
+func (c *memberContent) load(authEvents AuthEvents, userID string) error {
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: You have a `load` and `parse` function for this struct. That's confusing. You are consistently passing `authEvents` to all functions named `load`, so perhaps you should make this clearer e.g `loadFromAuthEvents` or to be more terse: `fromAuthEvents`.

File: eventauth/eventcontent.go [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+	memberEvent, err := authEvents.Member(userID)
+	if err != nil {
+		return err
+	}
+	if memberEvent == nil {
+		c.Membership = "leave"
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: This warrants a tiny explanatory comment to explain why.

File: eventauth/eventcontent.go [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+}
+
+// parse the power level content from an event.
+func (c *powerLevelContent) parse(event Event) error {
+	// Set the levels to their default values.
+	c.defaults()
+
+	// We can't extract the JSON directly to the powerLevelContent because we
+	// need to convert string values to int values.
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: This should probably also be mentioned in the struct definition. I nearly whined at you for it.

File: eventauth/eventcontent.go (hunk 4 above)
Comment (NegativeMjark) [inline eventauth/eventcontent.go]: Hmm, it's interesting. The auth code needs both top-level keys from the create event object, and the parsed create event content.
It turned out easier to write that code by shuffling the top-level keys into the same object as the create event content, rather than passing the two around separately.

File: eventauth/eventcontent.go (hunk 5 above)
Comment (NegativeMjark) [inline eventauth/eventcontent.go]: That is also true. I was documenting why I was extracting the key rather than repeating the documentation from the spec on where the key comes form. Do you think I linking to the spec would help?

File: eventauth/eventcontent.go (hunk 7 above)
Comment (NegativeMjark) [inline eventauth/eventcontent.go]: I was being purposely vague. But on closer inspection I'm only going to use this with user IDs so I might as well document it as such. 7c32ab3

File: eventauth/eventcontent.go (hunk 8 above)
Comment (NegativeMjark) [inline eventauth/eventcontent.go]: Done 7c32ab3

File: eventauth/eventcontent.go (hunk 11 above)
Comment (NegativeMjark) [inline eventauth/eventcontent.go]: Done d87fb92

File: eventauth/eventcontent.go (hunk 5 above)
Comment (Kegsay) [inline eventauth/eventcontent.go]: I wouldn't assume prior knowledge when it isn't self-explanatory - `creator` isn't clear if this is a user ID, server domain, or something completely different. A spec link would help clear up ambiguity and provide a reference for explaining what the field is for, so SGTM.

File: eventauth/eventcontent.go (hunk 4 above)
Comment (Kegsay) [inline eventauth/eventcontent.go]: I thought so, and yeah it's an interesting one. I think it can be confusing to pollute namespaces like this, but I understand that it may be even worse to have to pass around the same 2 structs to everything. Let's leave it for now and see how bad this gets as you add more logic.

File: eventauth/eventauth_test.go [hunk 12]
```diff
@@ -203,7 +203,7 @@ func testCaseJSON(t *testing.T, testCaseData string) {
 	for _, data := range tc.NotAllowed {
 		var event Event
 		if err := json.Unmarshal(data, &event); err != nil {
-			t.Fatal(err)
+			panic(err)
```
Comment (Kegsay) [inline eventauth/eventauth_test.go]: Any reasons for this? I thought `.Fatal` did panic?

File: eventauth/eventauth.go [hunk 13]
```diff
@@ -251,6 +229,73 @@ func redactEventAllowed(event Event, authEvents AuthEvents) error {
 	panic("Not implemented")
 }
 
+// defaultEventAllowed checks whether the event is allowed by the default
+// checks for events.
+// It returns an error if the event is not allowed or if there was a
+// problem loading the auth events needed.
```
Comment (Kegsay) [inline eventauth/eventauth.go]: Do we really want to bundle explicit event rejection into the same error response as "general problems loading the auth events"? I worry the latter may be transient, but the former isn't?

File: eventauth/eventauth.go [hunk 14]
```diff
@@ -251,6 +229,73 @@ func redactEventAllowed(event Event, authEvents AuthEvents) error {
...
+// An eventAllower has the information needed to authorise all events types
+// other than m.room.create, m.room.member and m.room.alias which are special.
+type eventAllower struct {
+	create      createContent
+	member      memberContent
+	powerLevels powerLevelContent
+}
+
+// setup loads the necesary information from the auth events.
+func (e *eventAllower) setup(authEvents AuthEvents, senderID string) error {
```
Comment (Kegsay) [inline eventauth/eventauth.go]: > necesary

Also, in general I thought "in order to set up this struct you need to execute this arbitrary code" stuff is typically done via a `New()` call which sets up the struct and returns the struct possibly with an error. Any reason for deviating from that convention here? Is it ever valid to be using an `eventAllower` *without* calling `setup()`?

File: eventauth/eventauth.go [hunk 15]
```diff
@@ -251,6 +229,73 @@ func redactEventAllowed(event Event, authEvents AuthEvents) error {
...
+	if err := e.create.load(authEvents); err != nil {
+		return err
+	}
+	if err := e.member.load(authEvents, senderID); err != nil {
+		return err
+	}
+	if err := e.powerLevels.load(authEvents, e.create.Creator); err != nil {
```
Comment (Kegsay) [inline eventauth/eventauth.go]: In a similar vein to https://github.com/matrix-org/gomatrixserverlib/pull/13/files#r95351964 - should we not have `New()` calls for these things as well?

File: eventauth/eventauth.go [hunk 16]
```diff
@@ -251,6 +229,73 @@ func redactEventAllowed(event Event, authEvents AuthEvents) error {
...
+		return err
+	}
+	return nil
+}
+
+// commonChecks does the checks that are applied to all events types other than
```
Comment (Kegsay) [inline eventauth/eventauth.go]: > events

File: eventauth/eventauth.go [hunk 17]
```diff
@@ -251,6 +229,73 @@ func redactEventAllowed(event Event, authEvents AuthEvents) error {
...
+	if event.RoomID != e.create.roomID {
+		return errorf("create event has different roomID: %q != %q", event.RoomID, e.create.roomID)
+	}
+
+	if err := e.create.userIDAllowed(event.Sender); err != nil {
+		return err
+	}
+
+	if e.member.Membership != "join" {
+		return errorf("sender %q not in room", event.Sender)
```
Comment (Kegsay) [inline eventauth/eventauth.go]: How does this work on transition states? E.g. invite -> join and join -> leave? Comment please.

File: eventauth/eventcontent.go [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+		return level
+	}
+	return c.userDefaultLevel
+}
+
+// eventLevel returns the power level needed to send an event in the room.
+func (c *powerLevelContent) eventLevel(eventType string, eventStateKey *string) int64 {
+	if eventType == "m.room.third_party_invite" {
+		// Special case third_party_invite events to have the same level as
+		// m.room.member invite events.
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: As per [the spec](http://matrix.org/docs/spec/client_server/r0.2.0.html#m-room-third-party-invite) - this event is a state event but you do not assert this is the case.

File: eventauth/eventcontent.go [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+	powerLevelsEvent, err := authEvents.PowerLevels()
+	if err != nil {
+		return err
+	}
+	if powerLevelsEvent != nil {
+		return c.parse(*powerLevelsEvent)
+	}
+
+	// If there are no power leves then fall back to defaults.
+	c.defaults()
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: Comment suggests there should be a conditional but there isn't one. Also:

> leves

File: eventauth/eventcontent_test.go [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+		if int64(want) != got.value {
+			t.Fatalf("Wanted %d got %q", want, got.value)
+		}
+	}
+}
+
+func TestLevelJSONValueInvalid(t *testing.T) {
+	var values []levelJSONValue
+	inputs := []string{
+		`[{}]`, `[[]]`, `["not a number"]`, `["0.0"]`,
```
Comment (Kegsay) [inline eventauth/eventcontent_test.go]: Oh, surprising! Why does 0.0 fail? Shouldn't that be 0? Is this for parity with python's `int()` impl?

File: eventauth/eventcontent.go [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+	if len(parts) != 2 {
+		// The ID must have a ":" character.
+		return "", errorf("invalid ID: %q", id)
+	}
+	// Return everything after the first ":" character.
+	return parts[1], nil
+}
+
+// memberContent is the JSON content of a m.room.member event needed for
+// for auth checks.
```
Comment (Kegsay) [inline eventauth/eventcontent.go]: > needed for for auth

File: eventauth/eventauth_test.go [hunk 22]
```diff
@@ -234,14 +236,282 @@ func TestAllowedEmptyRoom(t *testing.T) {
...
+		"allowed": [{
+			"type": "m.room.message",
+			"sender": "@u1:a",
+			"room_id": "!r1:a",
+			"event_id": "$e3:a",
+			"content": {"body": "Test"}
+		}],
+		"not_allowed": [{
+			"type": "m.room.message",
+			"sender": "@u2:b",
```
Comment (Kegsay) [inline eventauth/eventauth_test.go]: I'd prefer if you actually just varied the server rather than the local part as well here, so `@u1:b` because it's the domain part we care about.

File: eventauth/eventauth_test.go [hunk 23]
```diff
@@ -234,14 +236,282 @@ func TestAllowedEmptyRoom(t *testing.T) {
...
+			}
+		}, {
+			"type": "my.custom.state",
+			"state_key": "@u2:a",
+			"sender": "@u1:a",
+			"room_id": "!r1:a",
+			"event_id": "@e17:a",
+			"content": {},
+			"unsigned": {
+				"not_allowed": "State key starts with '@' and is for a different user"
```
Comment (Kegsay) [inline eventauth/eventauth_test.go]: :1st_place_medal: for consistently doing this on the tests, this makes them so nice to read ❤️

File: eventauth/eventcontent_test.go (hunk 20 above)
Comment (NegativeMjark) [inline eventauth/eventcontent_test.go]: "0.0" should fail because ```int("0.0")``` is not the same as ```int(float("0.0"))``` in python.

File: eventauth/eventauth_test.go (hunk 12 above)
Comment (NegativeMjark) [inline eventauth/eventauth_test.go]: panic gives me more interesting backtraces.

Comment (Kegsay) [inline eventauth/eventauth_test.go]: Fair enough!

File: eventauth/eventcontent_test.go (hunk 20 above)
Comment (Kegsay) [inline eventauth/eventcontent_test.go]: Thought so, okay then.

File: eventauth/eventcontent.go (hunk 6 above)
Comment (Kegsay) [inline eventauth/eventcontent.go]: Hmm, okay, maybe I just got confused then.

**Comment by NegativeMjark** [comment]
I think it would be better to hold off implementing the state_key checks for now since synapse doesn't perform any and I'd prefer that we kept the divergence from synapse's auth checks to a minimum.

**Comment by NegativeMjark** [comment]
Otherwise @Kegsay PTAL

**Comment by Kegsay** [comment]
> I think it would be better to hold off implementing the state_key checks for now since synapse doesn't perform any and I'd prefer that we kept the divergence from synapse's auth checks to a minimum.

That's fine, provided you put in enormous shouty caps that this is a known thing we're not addressing and the explanation as to why. Basically, what you've just said, but in comment form.

**Comment by Kegsay** [comment]
Other than that, LGTM.

**Comment by NegativeMjark** [comment]
Comment is added. @Kegsay are you happy?

**Comment by Kegsay** [comment]
LGTM
